### ISSUE TITLE
WebSocket connection failed

### ISSUE THREAD

**Issue by derrickb** [PepsRyuu/nollup#42]
So close to finally having nollup up and running with HMR! This is my first project where I've attempted to use HMR by the way, so I'm sure there are simple things I'm overlooking.

I'm calling nollup like this: `nollup -c --hot`

The bundle builds without errors and my app loads. But HMR isn't working and this is in the console:
```
WebSocket connection to 'ws://__hmr/' failed: Error during WebSocket handshake: Unexpected response code: 200
```
and the line that's throwing this error:
`var ws = new WebSocket('ws://' + window.location.host + '/__hmr');`

Any suggestions? Should I be filling something in for `--hmr-host`?

Thank you!

**Comment by derrickb** [comment]
Thanks for the help!
My app isn't Electron but something close; it's an Adobe Photoshop extension that runs in a modified nw.js... in other words, very similar, and likely the same advice for Electron applies here.

The WebSocket error is gone, but HMR still isn't working. I'm wondering if there's anything else I need for this to work. Here's some questions:

- My bundle file is accessed at `localhost:8080/build/dev/bundle.js`, so should my `--hmr-host` be `localhost:8080/build/dev`? Doesn't seem to be making a difference but thought I'd ask
- I am not loading the served version of index.html at `localhost:8080/build/dev/index.html`, do I need to be?

**Comment by PepsRyuu** [comment]
For ```--hmr-host```, it's just the hostname. No need to insert any paths. As long as the ```index.html``` file is loading the bundle it should be fine.

Do you see the ```/__hmr``` websocket in the network panel? Is it receiving any data when you save a file?

**Comment by derrickb** [comment]
Yes I see `/__hmr` under network, it says status `101`, size `0 B` and time `Pending`

if there's any other info you need, let me know!

**Comment by PepsRyuu** [comment]
Hmmm. If it's connecting to the HMR server, and it's not throwing any error on rebuild... I'm not entirely sure what it could be. As long as "--hot" is enabled, it should be pushes messages down the websocket. 

Are you able to reproduce this in a simple example?

**Comment by PepsRyuu** [comment]
Hey @derrickb, did you manage to figure out the issue?

**Comment by derrickb** [comment]
I'm afraid I've been very busy recently but will reply back as soon I can!
Still need to do a more barebones test.

**Comment by mikbry** [comment]
I am getting the same bug on FF and Chrome.
![Screenshot 2020-01-31 at 20 23 08](https://user-images.githubusercontent.com/395718/73567889-97f5a580-4467-11ea-8563-206c1179c2f6.png)
You see [hash] for filenames is not correctly setup. But hot reload for css is working.
Also I have the Terminal continuously Compilings files...

**Comment by PepsRyuu** [comment]
Thanks for the report @mikbry!

Do you have a reproducible example?

**Comment by mikbry** [comment]
Glad I could help ! See my PR above, it is a simple project, Where I try to have same structure as Create React App

**Comment by PepsRyuu** [comment]
Works on my end after the following changes:

* Commenting out the ```rollup-plugin-copy``` plugin. The reason it keeps rebuilding, is because Nollup is watching for file changes so it knows when to build, but because the copy plugin copies files on build, those changes are picked up, so Nollup builds again. It's an infinite loop. Not entirely sure what the best approach for this is yet. I might implement the Rollup watch ```include``` and ```exclude``` options for ```rollup.config.js```. Not sure it would be wise for Nollup to auto-detect whether or not it should ignore a file change. Suggestions on this are welcome.

* ``rollup-plugin-commonjs-alternate`` modified to allow for ```.jsx``` files. I'll need to publish an update for this to allow for various extensions. Will let you know when this is released.

* You were missing one final thing to get ```react-hot-loader``` working. In your ```App.jsx``` file, you need to include the following: 

```
if (process.env.NODE_ENV === 'development') {
    App = require('react-hot-loader').hot(module)(App);
}
```

**Comment by mikbry** [comment]
Thanks, I will try.
For the third point I understand. We need to bind hotReload to App as we did in early React app. Not a good point if we need to export without any change a Create React App project

**Comment by PepsRyuu** [comment]
Not sure I understand. Create React App doesn't include HMR by default from what I can tell, and the steps to include it are the same.

**Comment by PepsRyuu** [comment]
Released ```rollup-plugin-commonjs-alternate@0.2.0``` which includes the ```extensions``` option so it can be configured to recognise ```.jsx``` files. Looking into possible solutions for Nollup so that the workaround isn't needed.

**Comment by mikbry** [comment]
Create React App include HMR by default, I just tested it. I try to have the same code working on CRA and on Rollup/Nollup without any change.

**Comment by PepsRyuu** [comment]
Double-checked, ```create-react-app``` doesn't support "Hot Module Replacement", it supports "Hot Reloading". This is a different feature entirely, and unrelated to ```react-hot-loader``` which doesn't refresh the page but swap in components. 

I haven't implemented "Hot Reloading" into Nollup for two reasons: It's not always desirable to auto-refresh the page, and because HMR is the way forward for development and hot reloading would conflict in some situations. I left it entirely for the developer to decide if they want to incorporate the feature, which they can with a small snippet which can be either inserted into the code manual or through a plugin:

```
if (module) {
    module.hot.accept(() => window.location.reload());
}
```

On a side note, it's worth mentioning that React Refresh (the successor to React Hot Loader), doesn't require any code changes and provides Hot Module Replacement. You can see an example here: https://github.com/PepsRyuu/nollup/blob/master/examples/example-react-refresh/src/App.js

**Comment by mikbry** [comment]
Ok it works I updated to `rollup-plugin-commonjs-alternate@0.2.0`, removed copy and added manually hot function.
But It raised 2 points:
- There is a need for implemeting exclude include watch
- I got 'react-hot-loader react-🔥-dom patch is not detected' need to tweak a little more the code... So I will use React Refresh.

You are right `create-react-app` use by default Webpack LiveReload.

**Comment by PepsRyuu** [comment]
Looking into ```watch``` include and exclude options. No estimate yet on when it will be ready as it's going to require a fair bit of testing, but will update this thread when it's ready!

**Comment by mikbry** [comment]
Thanks, I could help if you need

**Comment by PepsRyuu** [comment]
Released ```0.10.4``` which implements Rollup ```watch.include/watch.exclude``` options.

**Comment by mikbry** [comment]
It works with `0.10.4` and watch.exclude no more infinite compiling 🎉
See my [repo](https://github.com/mikbry/RollupReactApp) I just made the PR.
